### PR TITLE
3978 Optimise products cache

### LINKED ISSUE #3978: French prod timeout issue: cannot add enterprise fees
## Description
Reported by users. When trying to add a fee, the snail appears with a 504 timeout on https://www.openfoodfrance.org/admin/enterprise_fees/bulk_update

## Steps to Reproduce
1. Log in as the Demo hub (ask me for access) on french prod
2. Go to `/admin/enterprise_fees?enterprise_id=212`
3. Try to add a fee and click "update"
4. See our friend the snail https://www.openfoodfrance.org/admin/enterprise_fees/bulk_update

## Animated Gif/Screenshot
image

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment
* Version used: v1.31
* Browser name and version: Firefox 67
* Operating System and version (desktop or mobile): Desktop

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #3978.

Updating enterprise fees became so slow that timeouts were reached and people couldn't change fees any more. I found this to be an issue between inefficient code in the ProductsCache and a very inefficient and almost unused override of `Exchange#eql?`. I approached the problem from two sides:

1. I made the ProductsCache more efficient, not using `exchanges.uniq` which is using `eql?`.
2. I removed the override of `eql?` to avoid similar mistakes in the future.

The increased performance affects updating fees and adding and removing products from order cycles.

#### What should we test?

Order cycles and product caching have been touched.

1. Activate products cache if not active on the current staging server (Settings / Caching).
1. Create an order cycle with multiple incoming and outgoing enterprises. Check the shopfront.
2. Change the products in the order cycle and check that they carry through to the shop.
2. Remove a producer from the order cycle and check the shop.
3. Change fees in the order cycle and make sure they carry through.

#### Release notes

Fixed: In rare circumstances a shopfront would not update after changing enterprise fees.
Fixed: Performance issues when updating enterprise fees.

Changelog Category: Fixed

### CODE REVIEW

File: lib/open_food_network/products_cache.rb [hunk 1]
```diff
@@ -170,7 +170,8 @@ def self.outgoing_exchanges_affected_by(exchanges)
           incoming_exchange.order_cycle,
           incoming_exchange.variant_ids
         )
-      end.flatten.uniq
+      end.flatten.uniq { |ex| [ex.receiver_id, ex.order_cycle_id]}
+      # Comparing only on these two ids is much faster.
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply lib/open_food_network/products_cache.rb]: Ah, okay, I'll do that next time. This code got replaced in the next commit anyway.

File: app/models/exchange.rb [hunk 2]
```diff
@@ -38,7 +38,7 @@ class Exchange < ActiveRecord::Base
   }
   scope :with_any_variant, lambda { |variant_ids|
     joins(:exchange_variants).
-      where('exchange_variants.variant_id IN (?)', variant_ids).
+      where(exchange_variants: { variant_id: variant_ids }).
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/exchange.rb]: Yes, I think so. The upside is that you can pass in ActiveRecord relations and they can be embedded in the same query.

I checked the Rails 4 upgrade guide and didn't see anything related to this. So I think we are good. I was wondering if we should set up some CI testing for Rails 4, moving that discussion to Slack.

**Comment by mkllnk** [comment]
Thank you @luisramos0. My favourite commits got squashed away, actually. I refactored the Exchange and then realised that the code can be removed. Sometimes refactoring is just learning and getting familiar with the code...

File: app/models/exchange.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kristinalim) [reply app/models/exchange.rb]: A gotcha is that chaining the scopes can override each other. I guess internally the hashes are merged.

```ruby
Spree::Variant.includes(:product).where(spree_products: {id: 1})
                                 .where(spree_products: {id: 2}).count
#    (1.4ms)  SELECT COUNT(DISTINCT "spree_variants"."id") FROM "spree_variants" LEFT OUTER JOIN "spree_products" ON "spree_products"."id" = "spree_variants"."product_id" WHERE "spree_products"."id" = 2 AND ("spree_variants".deleted_at IS NULL)
# => 2
Spree::Variant.includes(:product).where("spree_products.id = 1")
                                 .where("spree_products.id = 2").count
#    (1.5ms)  SELECT COUNT(DISTINCT "spree_variants"."id") FROM "spree_variants" LEFT OUTER JOIN "spree_products" ON "spree_products"."id" = "spree_variants"."product_id" WHERE ("spree_variants".deleted_at IS NULL) AND (spree_products.id = 1) AND (spree_products.id = 2)
# => 0
```

I like the hash form, but this is something to watch out for. They are not equivalent.

File: app/models/exchange.rb [hunk 3]
```diff
@@ -99,21 +99,6 @@ def participant
...
   end
 
-  def to_h(core_only = false)
-    h = attributes.merge('variant_ids' => variant_ids.sort,
-                         'enterprise_fee_ids' => enterprise_fee_ids.sort)
-    h.reject! { |k| %w(id order_cycle_id created_at updated_at).include? k } if core_only
-    h
-  end
-
-  def eql?(e)
```
Comment (kristinalim) [inline app/models/exchange.rb]: This is awesome. :fire: :joy: I'm sure this took a lot of courage @mkllnk.

File: lib/open_food_network/products_cache.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (luisramos0) [reply lib/open_food_network/products_cache.rb]: ah, sorry, I missed that the code didnt last long :-)

File: app/models/exchange.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (luisramos0) [reply app/models/exchange.rb]: very interesting @kristinalim

**Comment by RachL** [comment]
Thanks for the detailed tests! @mkllnk
I used the multishop already in staging
1. Change the products in the order cycle and check that they carry through to the shop. => this is looking good, I have change the prices, unit and name
 2. Removed Fred's farm and cheese and garlic where removed accordingly
3. Fees were applied accordingly (I applied the release testing on this)

File: app/models/exchange.rb (hunk 3 above)
Comment (mkllnk) [reply app/models/exchange.rb]: Yes, a little bit. How do you find out what code is affected by this change?

- There was no usage of `eql?`.
- There was no usage of `uniq` on anything that looked like exchanges.

`uniq` is mentioned in the Ruby documentation of `eql?`. So I was hoping that any other core method using `eql?` would be mentioned there as well. Hence, there should be no other code using it. :crossed_fingers:

**Pull request merged**